File: backend/src/main/java/org/jdk/project/service/config/RobotCommandService.java
package org.jdk.project.service.config;

import static org.jdk.project.service.config.ConfigCommandSupport.defaultInt;
import static org.jdk.project.service.config.ConfigCommandSupport.defaultString;
import static org.jdk.project.service.config.ConfigCommandSupport.ensureUpdated;
import static org.jdk.project.service.config.ConfigCommandSupport.requiredId;
import static org.jooq.generated.project.Tables.ROBOT;

import lombok.RequiredArgsConstructor;
import org.jdk.project.dto.config.RobotUpsertRequest;
import org.jooq.DSLContext;
import org.jooq.generated.project.tables.pojos.Robot;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class RobotCommandService {

  private final DSLContext dsl;

  @Transactional
  public Long create(RobotUpsertRequest request) {
    Robot robot = new Robot();
    robot.setLoungeId(requiredId(request.getLoungeId(), "贵宾室不能为空"));
    robot.setRegionId(request.getRegionId());
    robot.setRobotCode(request.getRobotId());
    robot.setName(request.getRobotName());
    robot.setMac(defaultString(request.getMac(), ""));
    robot.setIpAddress(defaultString(request.getRobotIp(), ""));
    robot.setRobotType(defaultString(request.getRobotType(), ""));
    robot.setBatteryPercent(defaultInt(request.getBatteryState(), 0));
    robot.setChargingState(defaultString(request.getChargingState(), ""));
    robot.setWorkingState(defaultString(request.getWorkingState(), ""));
    robot.setStandbyState(defaultString(request.getStandbyState(), ""));
    robot.setPositioningState(defaultString(request.getPositioningState(), ""));
    robot.setEnabled(request.getEnable() == null || request.getEnable() == 1);
    robot.setInitialCoordinate(defaultString(request.getOriCoordinate(), ""));
    robot.setAdminMode(Boolean.TRUE.equals(request.getAdminMode()));
    robot.setErrorCode(defaultString(request.getErrorCode(), ""));
    robot.setErrorMessage(defaultString(request.getErrorMessage(), ""));
    robot.setRemark(defaultString(request.getRemark(), ""));
    return dsl.insertInto(ROBOT)
        .set(dsl.newRecord(ROBOT, robot))
        .returningResult(ROBOT.ID)
        .fetchOne(ROBOT.ID);
  }

  @Transactional
  public void update(Long id, RobotUpsertRequest request) {
    ensureUpdated(
        dsl.update(ROBOT)
            .set(ROBOT.LOUNGE_ID, requiredId(request.getLoungeId(), "贵宾室不能为空"))
            .set(ROBOT.REGION_ID, request.getRegionId())
            .set(ROBOT.ROBOT_CODE, request.getRobotId())
            .set(ROBOT.NAME, request.getRobotName())
            .set(ROBOT.MAC, defaultString(request.getMac(), ""))
            .set(ROBOT.IP_ADDRESS, defaultString(request.getRobotIp(), ""))
            .set(ROBOT.ROBOT_TYPE, defaultString(request.getRobotType(), ""))
            .set(ROBOT.BATTERY_PERCENT, defaultInt(request.getBatteryState(), 0))
            .set(ROBOT.CHARGING_STATE, defaultString(request.getChargingState(), ""))
            .set(ROBOT.WORKING_STATE, defaultString(request.getWorkingState(), ""))
            .set(ROBOT.STANDBY_STATE, defaultString(request.getStandbyState(), ""))
            .set(ROBOT.POSITIONING_STATE, defaultString(request.getPositioningState(), ""))
            .set(ROBOT.ENABLED, request.getEnable() == null || request.getEnable() == 1)
            .set(ROBOT.INITIAL_COORDINATE, defaultString(request.getOriCoordinate(), ""))
            .set(ROBOT.ADMIN_MODE, Boolean.TRUE.equals(request.getAdminMode()))
            .set(ROBOT.ERROR_CODE, defaultString(request.getErrorCode(), ""))
            .set(ROBOT.ERROR_MESSAGE, defaultString(request.getErrorMessage(), ""))
            .set(ROBOT.REMARK, defaultString(request.getRemark(), ""))
            .where(ROBOT.ID.eq(id))
            .execute(),
        "机器人不存在");
  }

  @Transactional
  public void delete(Long id) {
    dsl.deleteFrom(ROBOT).where(ROBOT.ID.eq(id)).execute();
  }
}

File: backend/src/main/kotlin/org/jdk/project/service/config/RobotCommandService.kt
package org.jdk.project.service.config

import org.jdk.project.dto.config.RobotUpsertRequest
import org.jdk.project.service.config.ConfigCommandSupport.defaultInt
import org.jdk.project.service.config.ConfigCommandSupport.defaultString
import org.jdk.project.service.config.ConfigCommandSupport.ensureUpdated
import org.jdk.project.service.config.ConfigCommandSupport.requiredId
import org.jooq.DSLContext
import org.jooq.generated.project.Tables.ROBOT
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RobotCommandService(
    private val dsl: DSLContext,
) {
    @Transactional
    fun create(request: RobotUpsertRequest): Long? =
        dsl.insertInto(ROBOT)
            .set(ROBOT.LOUNGE_ID, requiredId(request.loungeId, "贵宾室不能为空"))
            .set(ROBOT.REGION_ID, request.regionId)
            .set(ROBOT.ROBOT_CODE, request.robotId)
            .set(ROBOT.NAME, request.robotName)
            .set(ROBOT.MAC, defaultString(request.mac, ""))
            .set(ROBOT.IP_ADDRESS, defaultString(request.robotIp, ""))
            .set(ROBOT.ROBOT_TYPE, defaultString(request.robotType, ""))
            .set(ROBOT.BATTERY_PERCENT, defaultInt(request.batteryState, 0))
            .set(ROBOT.CHARGING_STATE, defaultString(request.chargingState, ""))
            .set(ROBOT.WORKING_STATE, defaultString(request.workingState, ""))
            .set(ROBOT.STANDBY_STATE, defaultString(request.standbyState, ""))
            .set(ROBOT.POSITIONING_STATE, defaultString(request.positioningState, ""))
            .set(ROBOT.ENABLED, request.enable == null || request.enable == 1)
            .set(ROBOT.INITIAL_COORDINATE, defaultString(request.oriCoordinate, ""))
            .set(ROBOT.ADMIN_MODE, request.adminMode == true)
            .set(ROBOT.ERROR_CODE, defaultString(request.errorCode, ""))
            .set(ROBOT.ERROR_MESSAGE, defaultString(request.errorMessage, ""))
            .set(ROBOT.REMARK, defaultString(request.remark, ""))
            .returningResult(ROBOT.ID)
            .fetchOne(ROBOT.ID)

    @Transactional
    fun update(id: Long, request: RobotUpsertRequest) {
        ensureUpdated(
            dsl.update(ROBOT)
                .set(ROBOT.LOUNGE_ID, requiredId(request.loungeId, "贵宾室不能为空"))
                .set(ROBOT.REGION_ID, request.regionId)
                .set(ROBOT.ROBOT_CODE, request.robotId)
                .set(ROBOT.NAME, request.robotName)
                .set(ROBOT.MAC, defaultString(request.mac, ""))
                .set(ROBOT.IP_ADDRESS, defaultString(request.robotIp, ""))
                .set(ROBOT.ROBOT_TYPE, defaultString(request.robotType, ""))
                .set(ROBOT.BATTERY_PERCENT, defaultInt(request.batteryState, 0))
                .set(ROBOT.CHARGING_STATE, defaultString(request.chargingState, ""))
                .set(ROBOT.WORKING_STATE, defaultString(request.workingState, ""))
                .set(ROBOT.STANDBY_STATE, defaultString(request.standbyState, ""))
                .set(ROBOT.POSITIONING_STATE, defaultString(request.positioningState, ""))
                .set(ROBOT.ENABLED, request.enable == null || request.enable == 1)
                .set(ROBOT.INITIAL_COORDINATE, defaultString(request.oriCoordinate, ""))
                .set(ROBOT.ADMIN_MODE, request.adminMode == true)
                .set(ROBOT.ERROR_CODE, defaultString(request.errorCode, ""))
                .set(ROBOT.ERROR_MESSAGE, defaultString(request.errorMessage, ""))
                .set(ROBOT.REMARK, defaultString(request.remark, ""))
                .where(ROBOT.ID.eq(id))
                .execute(),
            "机器人不存在",
        )
    }

    @Transactional
    fun delete(id: Long) {
        dsl.deleteFrom(ROBOT).where(ROBOT.ID.eq(id)).execute()
    }
}
